include/oneapi/dpl/internal/reduce_by_segment_impl.h

```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+                        __out_values[__idx] = __binary_op(__carry_in, __loc_partials[__i - __start]);
+                        __apply_aggs = false;
+                    }
+                    else
+                        __out_values[__idx] = __loc_partials[__i - __start];
+
+                    __out_keys[__idx] = __keys[__i]; 
+                    ++__item_offset;
+                }
+            }
+
+            // 2e. Output the work group aggregate and total number of segments for use in phase 2.
+            if (__local_id == __wgroup_size - 1) // last work item writes the group's carry out
+            {
+                // If no segment ends in the item, the aggregates from previous work groups must be applied.
+                if (__max_end == 0)
+                    __partials_acc[__group_id] =
+                        __binary_op(__carry_in, __accumulator); // needs to be inclusive with last element
+
+                else
+                    __partials_acc[__group_id] = __accumulator;
+            }
+        });
+    });
+
+    // 3. Apply inter work-group aggregates
+    __exec.queue()
+        .submit([&](sycl::handler& __cgh) {
+            oneapi::dpl::__ranges::__require_access(__cgh, __keys, __out_keys, __out_values);
+
+            auto __partials_acc = __partials.template get_access<sycl::access_mode::read>(__cgh);
+            auto __seg_ends_acc = __seg_ends.template get_access<sycl::access_mode::read>(__cgh);
+            auto __end_idx_acc = __end_idx.template get_access<sycl::access_mode::write>(__cgh);
+
+            __cgh.depends_on(__wg_reduce);
+
+            __cgh.parallel_for(sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1>
+                                                                                                     __item) {
+                auto __group = __item.get_group();
+                auto __group_id = __group.get_group_id(0);
+                auto __global_id = __item.get_global_id();
+                auto __local_id = __item.get_local_id();
+
+                int32_t __start = __global_id * __vals_per_item;
+                int32_t __end = sycl::minimum{}(__start + __vals_per_item, __n);
+                int32_t __item_segments = 0;
+
+                int32_t __wg_agg_idx = __group_id - 1;
+                __val_type __agg_collector{};
+
+                // 3a. Check to see if an aggregate exists and compute that value in the first 
+                // work item.
+                bool __first = true;
+                bool __ag_exists = false;
+                if (__local_id == 0 && __wg_agg_idx >= 0)
+                {
+                    if (__start < __n && __keys[__start] == __keys[__start - 1])
+                    {
+                        __ag_exists = true;
+                        for (int32_t __i = __wg_agg_idx; __i >= 0; --__i)
+                        {
+                            const auto& __wg_aggregate = __partials_acc[__i];
```

Comment:
Is it really required to save this and next values by const reference? May be better simple copy them into local const variable?

Reply:
The type of `__partials` is unknown, so I think it may be best to save by const reference since we do not know if the copy is expensive. However, the next line accesses a boolean value, so using a copy into a local const variable I agree is best.

Reply:
This is now out of date, we are just using these accessor values directly in the following lines rather than grabbing a copy / reference and using that.